archunit/src/main/java/com/tngtech/archunit/library/cycle_detection/rules/CycleArchCondition.java

```diff
@@ -0,0 +1,382 @@
+        private final CycleRuleConfiguration cycleConfiguration = new CycleRuleConfiguration();
+        private final Function<COMPONENT, String> getDescriptionOfComponent;
+
+        private EventRecorder(Function<COMPONENT, String> getDescriptionOfComponent) {
+            this.getDescriptionOfComponent = getDescriptionOfComponent;
+            log.trace("Maximum number of dependencies to report per edge is set to {}; "
+                            + "this limit can be adapted using the `archunit.properties` value `{}=xxx`",
+                    cycleConfiguration.getMaxNumberOfDependenciesToShowPerEdge(), MAX_NUMBER_OF_DEPENDENCIES_TO_SHOW_PER_EDGE_PROPERTY_NAME);
+        }
+
+        void record(Cycle<ComponentDependency<COMPONENT>> cycle, ConditionEvents events) {
+            events.add(newEvent(cycle));
+        }
+
+        private ConditionEvent newEvent(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            Map<String, ComponentDependency<COMPONENT>> descriptionsToEdges = sortEdgesByDescription(cycle);
+            String description = createDescription(descriptionsToEdges.keySet(), CYCLE_DETECTED_SECTION_INTRO.length());
+            String details = createDetails(descriptionsToEdges);
+            return new SimpleConditionEvent(cycle,
+                    false,
+                    CYCLE_DETECTED_SECTION_INTRO + description + lineSeparator() + details);
+        }
+
+        private Map<String, ComponentDependency<COMPONENT>> sortEdgesByDescription(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            LinkedList<ComponentDependency<COMPONENT>> edges = new LinkedList<>(cycle.getEdges());
+            ComponentDependency<COMPONENT> startEdge = findStartEdge(cycle);
+            while (!edges.getFirst().equals(startEdge)) {
+                edges.addLast(edges.pollFirst());
+            }
+            Map<String, ComponentDependency<COMPONENT>> descriptionToEdge = new LinkedHashMap<>();
+            for (ComponentDependency<COMPONENT> edge : edges) {
+                descriptionToEdge.put(getDescriptionOfComponent.apply(edge.getOrigin()), edge);
+            }
+            return descriptionToEdge;
+        }
+
+        // A cycle always has edges, so we know that there is always at least one edge and by that a minimum element
+        // with respect to comparing the description lexicographically
+        @SuppressWarnings("OptionalGetWithoutIsPresent")
+        private ComponentDependency<COMPONENT> findStartEdge(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            return cycle.getEdges().stream().min(comparing(input -> getDescriptionOfComponent.apply(input.getOrigin()))).get();
+        }
+
+        private String createDescription(Collection<String> edgeDescriptions, int indent) {
+            List<String> descriptions = new ArrayList<>(edgeDescriptions);
+            descriptions.add(descriptions.get(0));
+            return Joiner.on(" -> " + lineSeparator() + Strings.repeat(" ", indent)).join(descriptions);
```

Comment:
`lineSeparator() + Strings.repeat(" ", CYCLE_DETECTED_SECTION_INTRO.length())` could be a static constant (`CYCLE_SEPARATOR`). Currently, it is re-calculated for every cycle.

I assume that you wanted to have both usages of `CYCLE_DETECTED_SECTION_INTRO` next to each other in `newEvent`, but IMO it's not obvious how `CYCLE_DETECTED_SECTION_INTRO.length()` is used in `createDescription` anyways.

Reply:
Hmm, yeah, the idea was to see the connection at the call site :thinking: But you're right, defining constants next to each other is likely also good enough...